x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java

```diff
@@ -0,0 +1,171 @@
+
+    private final NodeClient nodeClient;
+    private final TransportService transportService;
+
+    @Inject
+    public TransportGetTopNFunctionsAction(NodeClient nodeClient, TransportService transportService, ActionFilters actionFilters) {
+        super(
+            GetTopNFunctionsAction.NAME,
+            transportService,
+            actionFilters,
+            GetStackTracesRequest::new,
+            EsExecutors.DIRECT_EXECUTOR_SERVICE
+        );
+        this.nodeClient = nodeClient;
+        this.transportService = transportService;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetStackTracesRequest request, ActionListener<GetTopNFunctionsResponse> listener) {
+        Client client = new ParentTaskAssigningClient(this.nodeClient, transportService.getLocalNode(), task);
+        StopWatch watch = new StopWatch("getTopNFunctionsAction");
+        client.execute(GetStackTracesAction.INSTANCE, request, ActionListener.wrap(searchResponse -> {
+            long responseStart = System.nanoTime();
+            StopWatch processingWatch = new StopWatch("Processing response");
+            GetTopNFunctionsResponse topNFunctionsResponse = buildTopNFunctions(searchResponse);
+            log.debug(() -> watch.report() + " " + processingWatch.report());
+            listener.onResponse(topNFunctionsResponse);
+        }, listener::onFailure));
+    }
+
+    static GetTopNFunctionsResponse buildTopNFunctions(GetStackTracesResponse response) {
+        TopNFunctionsBuilder builder = new TopNFunctionsBuilder(response.getSamplingRate());
+        if (response.getTotalFrames() == 0) {
+            return builder.build();
+        }
+
+        for (Map.Entry<String, StackTrace> st : response.getStackTraces().entrySet()) {
+            Set<String> frameGroupsPerStackTrace = new HashSet<String>();
+            String stackTraceId = st.getKey();
+            StackTrace stackTrace = st.getValue();
+            long samples = response.getStackTraceEvents().getOrDefault(stackTraceId, 0L);
+            builder.addTotalCount(samples);
+
+            int frameCount = stackTrace.frameIds.size();
+            for (int i = 0; i < frameCount; i++) {
+                String frameId = stackTrace.frameIds.get(i);
+                String fileId = stackTrace.fileIds.get(i);
+                Integer frameType = stackTrace.typeIds.get(i);
+                Integer addressOrLine = stackTrace.addressOrLines.get(i);
+                StackFrame stackFrame = response.getStackFrames().getOrDefault(frameId, EMPTY_STACKFRAME);
+                String executable = response.getExecutables().getOrDefault(fileId, "");
+
+                for (Frame frame : stackFrame.frames()) {
+                    String frameGroupId = FrameGroupID.create(fileId, addressOrLine, executable, frame.fileName(), frame.functionName());
+                    if (builder.setCurrentTopNFunction(frameGroupId) == false) {
```

Comment:
It's confusing to me that a setter returns a boolean and I constantly need to look at the implementation to understand what it means. Can we change the implementation so it's more apparent what's happening. E.g. if I look at similar code in `TransportGetFlamegraphAction` I have a much better grasp of what's happening on a high level.